cni/pkg/ebpf/app/ambient_redirect.bpf.c

```diff
@@ -0,0 +1,277 @@
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP && iph->protocol != IPPROTO_UDP)
+        return TC_ACT_OK;
+
+    pi = get_app_info_from_ipv4(iph->daddr);
+    if (!pi) {
+        // ip is not in ambient managed mesh
+        // dbg("ztunnel_host_ingress unable to retrieve pod info: 0x%x\n", bpf_ntohl(iph->daddr));
+        return TC_ACT_OK;
+    }
+
+    __builtin_memcpy(eth->h_dest, pi->mac_addr, ETH_ALEN);
+    skb->cb[4] = BYPASS_CB;
+
+    dbg("ztunnel redirect to app(0x%x) ifidx: %u\n", bpf_ntohl(iph->daddr), pi->ifindex);
+
+    return bpf_redirect(pi->ifindex, 0);
+}
+
+// For ztunnel pod veth pair(in pod ns) ingress hook
+SEC("tc")
+int ztunnel_ingress(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    struct ethhdr  *eth = data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct bpf_sock_tuple *tuple;
+    size_t tuple_len;
+    struct bpf_sock *sk;
+    int skip_mark = 0;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    struct iphdr *iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support UDP tunneling
+    if (iph->protocol != IPPROTO_TCP)
+        return TC_ACT_OK;
+
+    tuple = (struct bpf_sock_tuple *)&iph->saddr;
+    tuple_len = sizeof(tuple->ipv4);
+    if ((void *)tuple + sizeof(tuple->ipv4) > (void *)(long)skb->data_end)
+        return TC_ACT_SHOT;
+
+    if (skb->cb[4] == OUTBOUND_CB) {
+        // We mark all app egress pkt as outbound and redirect here.
+        // We need identify if it's a actual *outbound* or just a reponse to
+        // the proxy
+        sk = bpf_skc_lookup_tcp(skb, tuple, tuple_len, BPF_F_CURRENT_NETNS, 0);
+        if (sk) {
+            if (sk->state != BPF_TCP_LISTEN) {
+                skip_mark = 1;
+            }
+            bpf_sk_release(sk);
+        }
+        if (!skip_mark) {
+            skb->mark = ZTUNNEL_OUTBOUND_MARK;
+            dbg("got packet with out cb mark 0x%x, mark packet: %u\n", skb->cb[4], skb->mark);
+        }
+    } else if (skb->cb[4] == INBOUND_CB) {
+        // For original source case, app will overwrite
+        // the cb[4] and redirect here.
+        sk = bpf_skc_lookup_tcp(skb, tuple, tuple_len, BPF_F_CURRENT_NETNS, 0);
+        if (sk) {
+            if (sk->state != BPF_TCP_LISTEN
+                && sk->src_port != ZTUNNEL_INBOUND_PORT
+                && sk->src_port != ZTUNNEL_INBOUND_PLAINTEXT_PORT) {
+                skip_mark = 1;
+            }
+            bpf_sk_release(sk);
+        }
+        if (!skip_mark) {
+            skb->mark = ZTUNNEL_INBOUND_MARK;
+            dbg("got inbound packet with in cb 0x%x, mark packet: %u\n", skb->cb[4], skb->mark);
+        }
+    }
+
+    return TC_ACT_OK;
+}
+
+char __license[] SEC("license") = "Dual BSD/GPL";
```

Comment:
I might be misinterpreting this.. but the Istio project is under the Apache 2.0 license and that's also in the header of this file. Curious why this is here stating a license of Dual BSD/GPL?

Reply:
this is the kernel license requirement

Reply:
@howardjohn @ericvn can you check with your lawyer on license? GPL is not open source friendly afaik. Also, does this mean we need to add GPL to Istio's license?

Reply:
It's a good call, we need to be careful with it: https://github.com/nyrahul/ebpf-guide/blob/master/docs/gpl_license_ebpf.rst

Reply:
yep, this is the current approach: ebpf is dual license and user space code is apache.